math/primes.go

```diff
@@ -0,0 +1,32 @@
+package math
+
+import (
+	"crypto/rand"
+	"io"
+	"math/big"
+)
+
+// IsSafePrime reports whether p is (probably) a safe prime.
+// The prime p=2*q+1 is safe prime if both p and q are primes.
+// The n parameter is the number of checks for primality test.
+func IsSafePrime(p *big.Int, n int) bool {
+	pdiv2 := new(big.Int).Rsh(p, 1)
+	return p.ProbablyPrime(n) && pdiv2.ProbablyPrime(n)
+}
+
+// SafePrime returns a number of the given bit length that is a safe prime with high probability.
+// The number returned p=2*q+1 is a safe prime if both p and q are primes.
+// SafePrime will return error for any error returned by rand.Read or if bits < 2.
+func SafePrime(random io.Reader, bits int) (*big.Int, error) {
```

Comment:
I think it's faster to find a prime `q`, and then check whether `2q+1` is prime.

Reply:
true, changed